File: Kotlin/프로그래머스/1/64061. 크레인 인형뽑기 게임/크레인 인형뽑기 게임.kt
class Solution {
    lateinit var globalBoard: Array<IntArray>
    
    fun solution(board: Array<IntArray>, moves: IntArray): Int {
        var answer = 0
        globalBoard = board
        // 뽑은 인형을 넣는 바구니
        val basket = mutableListOf<Int>()
        
        for (move in moves) {
            val doll = get_poppable_doll_of_col(move - 1)
            
            // 뽑은 인형이 없는 경우(빈 줄) 무시
            if (doll == 0) continue
        
            // 바구니가 비어있지 않고 
            if (basket.isNotEmpty() && basket.last() == doll) { // 맨 위 인형과 일치하는 경우
                basket.removeLast() // 기존 인형 제거
                answer += 2 // 인형 2개가 사라지므로 +2
            } 
            else { // 맨 위 인형과 일치하지 않는 경우
                basket.add(doll)
            }
        }
        
        return answer
    }
    
    // 세로줄 가장 위의 인형 구하기
    fun get_poppable_doll_of_col(col: Int): Int {
        for (row in globalBoard.indices) {
            if (globalBoard[row][col] != 0) {
                val top_doll = globalBoard[row][col]
                globalBoard[row][col] = 0
                return top_doll
            }
        }
        return 0 // 해당 줄이 모두 비어있으면 0 반환
    }
}

File: Python/프로그래머스/1/64061. 크레인 인형뽑기 게임/크레인 인형뽑기 게임.py
def solution(board, moves):
    answer = 0
    basket = []

    # 세로줄 가장 위의 인형 구하기
    def get_poppable_doll_of_col(col):
        for line in board:
            if line[col] != 0:
                top_doll = line[col]
                line[col] = 0
                return top_doll
        return None # 해당 줄이 모두 비어있으면 None 반환

    # 크레인 작업 시작
    for move in moves:
        doll = get_poppable_doll_of_col(move - 1)
        
        # 뽑은 인형이 없는 경우(빈 줄) 무시
        if doll is None:
            continue
            
        # 바구니가 비어있지 않고, 맨 위 인형과 일치하는 경우
        if basket and basket[-1] == doll:
            basket.pop() # 기존 인형 제거
            answer += 2  # 인형 2개가 사라지므로 +2
        else:
            basket.append(doll) # 바구니에 추가

    return answer
